pkg/autoscaler/autoscaler.go

```diff
@@ -732,6 +518,96 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
 			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
 		})
 
+		It("scales up and down while respecting MaxNodesTotal [Slow][Serial]", func() {
+			By("Creating 1 MachineSet with 1 replica")
+			var transientMachineSet *mapiv1beta1.MachineSet
+			targetedNodeLabel := fmt.Sprintf("%v-scale-updown", autoscalerWorkerNodeRoleLabel)
+			machineSetParams := framework.BuildMachineSetParams(client, 1)
+			machineSetParams.Labels[targetedNodeLabel] = ""
+			transientMachineSet, err = framework.CreateMachineSet(client, machineSetParams)
+			Expect(err).ToNot(HaveOccurred())
+			cleanupObjects[transientMachineSet.GetName()] = transientMachineSet
+
+			By("Waiting for all Machines in the MachineSet to enter Running phase")
+			framework.WaitForMachineSet(client, transientMachineSet.GetName())
+
+			// To exercise the MaxNodesTotal mechanism we want to craft this
+			// MachineAutoscaler so that its maximum size would cause the cluster
+			// to grow beyond the MaxNodesTotal. This is calculated as the
+			// MaxNodesTotal - current number of nodes
+			// + 1 (for the initial size of the MachineSet)
+			// + 1 (to exceed the maximum)
+			By("Calculating maximum size for MachineAutoscaler")
+			nodes, err := framework.GetNodes(client)
+			Expect(err).NotTo(HaveOccurred())
+			maxMachineSetReplicas := int32(caMaxNodesTotal - len(nodes) + 2)
```

Comment:
Why not just have the `maxMachineSetReplicas` as `caMaxNodesTotal + 1`, then the maximum size of this machineset is definitely bigger than the max size of the cluster and we don't need to worry about this sum? We are never going to hit the max right? (Assuming everything works 😅 )

Reply:
i thought about that, i think i went with this because if everything works as we expect it these sizes should hold accurate. i can certainly switch to just use the max cluster size though.